gateway/ln-gateway/src/actor_fork.rs

```diff
@@ -44,7 +55,112 @@ impl GatewayActor {
             Err(e) => warn!("Failed to connect with federation: {}", e),
         }
 
-        Ok(Self { client })
+        let actor = Self {
+            client,
+            lnrpc,
+            task_group,
+        };
+
+        actor.subscribe_htlcs().await?;
+
+        Ok(actor)
+    }
+
+    async fn subscribe_htlcs(&self) -> Result<()> {
+        let actor = self.to_owned();
+        let lnrpc = self.lnrpc.to_owned();
+        let short_channel_id = self.client.config().mint_channel_id;
+        let mut tg = self.task_group.clone();
+
+        let mut stream = lnrpc
+            .subscribe_intercept_htlcs(SubscribeInterceptHtlcsRequest { short_channel_id })
+            .await?;
+
+        tg.spawn(
+            "Subscribe to intercepted HTLCs in stream",
+            move |subscription| async move {
+                loop {
+                    if subscription.is_shutting_down() {
+                        info!("Shutting down HTLC handler");
+                        // TODO: Unsubscribe from HTLCs?
+                        break;
+                    }
+
+                    let mut htlc_outcomes = Vec::<CompleteHtlcsRequest>::new();
+
+                    while let Some(SubscribeInterceptHtlcsResponse {
+                        payment_hash,
+                        outgoing_amount_msat,
+                        intercepted_htlc_id,
+                        ..
+                    }) = match stream.message().await {
+                        Ok(Some(msg)) => Some(msg),
+                        Ok(None) => {
+                            warn!("Stream closed");
+                            None
+                        }
```

Comment:
These nested loops don't make any sense. If the stream closes the outer `loop` will lead to an infinite loop using 100% CPU without re-establishing the stream. As long as the stream is active the outer `loop` is not needed.

Reply:
I was quite convinced we needed both loops. The outer one for keeping a persistent and long running subscription for htlcs. The inner While loop for reading htlc messages from the stream. Probably carried over these notions from working with plain mpsc channel subscriptions

[pending, will fix]